src/frequenz/sdk/data_ingestion/resampling/timeseries_resampling_actor.py

```diff
@@ -0,0 +1,159 @@
+from frequenz.sdk.data_ingestion.resampling.subscriptions import (
+    Subscription,
+    SubscriptionRequest,
+    SubscriptionResponse,
+)
+from frequenz.sdk.data_ingestion.resampling.timeseries_resampler import (
+    TimeSeriesResampler,
+)
+
+logger = logging.Logger(__name__)
+
+ResamplingFunction = Callable[[Sequence[Sample], float], float]
+
+
+# pylint: disable=unused-argument
+def average(samples: Sequence[Sample], resampling_frequency: float) -> float:
+    """Calculate average of the provided values.
+
+    Args:
+        samples: sequences of samples to apply the average to
+        resampling_frequency: resampling frequency, in seconds
+
+    Returns:
+        average of all the sample values
+    """
+    if len(samples) == 0:
+        return 0.0
+    return sum(sample.value for sample in samples) / len(samples)
+
+
+@actor
+class TimeSeriesResamplingActor:
+    """TimeSeriesResamplingActor used to ingest component data and resample it."""
+
+    def __init__(  # pylint: disable=too-many-arguments
+        self,
+        data_provider: DataProvider,
+        subscription_receiver: Receiver[SubscriptionRequest],
+        period_s: float = 0.2,
+        max_data_age_in_periods: float = 3.0,
+        resampling_function: ResamplingFunction = average,
+    ) -> None:
+        """Initialize the TimeSeriesResamplingActor.
+
+        Args:
+            data_provider: class implementing the DataProvider interface for
+                getting raw component data
+            subscription_receiver: channel for receiving subscription requests
+            period_s: resampling frequency, in seconds
+            max_data_age_in_periods: max age that samples shouldn't exceed in order
+                to be used in the resampling function
+            resampling_function: function to be applied to a sequence of samples within
+                a resampling period to produce a single output sample
+        """
+        self._data_provider = data_provider
+        self._subscription_receiver = subscription_receiver
+        self._period_s = period_s
+        self._max_data_age_in_periods: float = max_data_age_in_periods
+        self._resampling_function: ResamplingFunction = resampling_function
+
+        self._resampler = TimeSeriesResampler(
+            period_s=period_s,
+            max_data_age_in_periods=max_data_age_in_periods,
+            resampling_function=resampling_function,
+        )
+
+        # `TimeSeriesId`s as dictionary keys will likely have to change to
+        # `Subscription`s, which take `start_time` into account for reading
+        # historical data
+        self._input_receivers: Dict[TimeSeriesId, Receiver[Optional[Sample]]] = {}
+        self._output_channels: Dict[TimeSeriesId, Broadcast[Optional[Sample]]] = {}
+        self._resampling_timer = Timer(interval=self._period_s)
+
+    def subscribe(self, subscription: Subscription) -> Receiver[Optional[Sample]]:
```

Comment:
Anyway, the subscription will very likely change, so let's not touch this for now.

Reply:
> I think a receiver data can't be `Optional` because the channel implementation uses `None` as a special value to mark the end of the stream (I would like to change that in the future, but I think for now we can´t do it).

`Optional[Sample]` is no longer needed and I will change it to `Sample`, because the sample's value can now be `None` as discussed in:
https://github.com/frequenz-floss/frequenz-sdk-python/pull/32#discussion_r1000209375

Reply:
> Just taste, I would just use the arguments here directly without going through a `Subscription` object.

What do you think about leaving `Subscription` and `TimeSeriesId` as they are now and adjusting them once `DataSourcingActor` is ready?

Basically the `TimeSeriesId` is `(component_id, metric_id)` and `Subscription` is `(component_id, metric_id, start_time)`, but we can rename these once it is clear what names should be used.

Reply:
> Basically the TimeSeriesId is (component_id, metric_id)

This for sure is wrong, TimeseriesId is a `str`. The subscription, that will probably go away (or just be part of the data sourcing actor), can be as you say.

But yeah, for now it is fine not to touch this part until we confirm all of this.

But for more context, please have a look at https://github.com/frequenz-floss/frequenz-sdk-python/issues/30#issuecomment-1283925616.

Reply:
For now we are finally coupling the data source actor and the resampler, so there will be no opaque `timeseries_id`, the resampler will know about `component_id` and `metric_id` as in the original design in the issue. So we can mark this as resolved for now.